posts/inside-rust/2024-09-23-rtn-call-for-testing.md

```diff
@@ -0,0 +1,165 @@
+---
+layout: post
+title: "Return Type Notation MVP: Call for Testing!"
+author: Michael Goulet
+team: The Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
+---
+
+The async working group is excited to announce that [RFC 3654] return type notation (RTN) is ready for testing. In this post, we want to briefly describe the feature and announce a call for testing on nightly Rust.
+
+## The backstory
+
+Rust 1.75 [stabilized](https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html) async fn in traits (AFIT) and return-position impl Trait in traits (RPITIT). These desugar to anonymous generic associated types (GATs). However, unlike GATs, users of these types cannot use `where` clauses to further restrict these return types. This is known as the ["send bound"](https://smallcultfollowing.com/babysteps/blog/2023/02/01/async-trait-send-bounds-part-1-intro/) problem, since it often affects `Send` bounds on futures in the async ecosystem.
+
+### An example
+
+Consider a trait `Foo` with a `method` that returns a type of `impl Future<Output = ()>`. We want to write a function that calls `method` and spawns the future on another thread:
+
+```rust
+fn spawn<T>(f: impl Future<Output = T> + Send + 'static) {}
+
+trait Foo {
+    fn method() -> impl Future<Output = ()>; // <-- RPITIT.
+}
+
+fn needs_sendable_future<T: Foo>()
+where
+    // How do we further restrict `T::method()` to be `Send + 'static`?
+{
+    spawn(T::method());
+    //~^ ERROR: `impl Future<Output = ()>` is not `Send`!
+}
+```
+
+Specifically, we may not want to restrict the *declaration* of `Foo`, since changing it in the declaration would restrict *all* implementations of `Foo`.
+
+```rust
+trait Foo {
+    fn method() -> impl Future<Output = ()> + Send + 'static;
+    //                                      ~~~~~~~~~~~~~~~~
+    //                                      Not what we want.
+}
+```
+
+So, on stable Rust, we have no way of expressing this restriction when using AFIT or RPITIT. In contrast, we can express this today if we were to use a GAT directly:
```

Comment:
```suggestion
So, on stable Rust, we have no way of expressing this restriction when using async functions in traits or RPITIT. In contrast, we can express this today if we were to use a generic associated type (GAT) directly:
```

Reply:
Why did you change AFIT but not RPITIT? All of these acronyms are spelled out in the backstory section; we could either consistently use acronyms or the full names, but this seems a bit inconsistent... Your edit is literally doing all three of:
* Only full name
* Only acronym
* Both full name and acronym

Reply:
It's inconsistent (and wrong because all these are spelled out in the backstory section), but my thinking was that I'd only seen AFIT used once so we might as well spell it out and not bother with the acronym, RPITIT had already been used, and GAT was used later.

Anyway, this suggestion doesn't make sense so I withdraw it!